### ISSUE TITLE
tensorflow implement D2D

### ISSUE THREAD

**Issue by tom99763** [POSTECH-CVLab/PyTorch-StudioGAN#137]
Is that right?
`
class Reboot_Softmax(layers.Layer):

    def __init__(self, nclass, tau=0.5, mp=0.95):
        super().__init__()
        self.nclass = nclass
        self.tau = tau
        self.mp = mp

    def build(self, shape):
        f_shape, y_shape = shape
        self.embMap = layers.Embedding(self.nclass, f_shape[-1])

    def call(self, inputs):
        f, y = inputs
        v = self.embMap(y)
        y = tf.one_hot(y, self.nclass)
        # normalize
        f = f / tf.norm(f, axis=-1, keepdims=True)
        v = v / tf.norm(v, axis=-1, keepdims=True)
        # D2D
        D2D_sim = f @ tf.transpose(f)  # NxN
        mask = y @ tf.transpose(y)  # NxN
        D2D_sim = D2D_sim * (1. - mask)  # NxN false negative mask
        D2D_sim = tf.math.reduce_sum(tf.exp(tf.math.maximum(D2D_sim - (1 - self.mp), 0.) / self.tau), axis=-1)  # N
        # D2C
        D2C_sim = tf.linalg.diag_part(f @ tf.transpose(v))  # N
        D2C_sim = tf.exp(tf.math.minimum(D2C_sim - self.mp, 0.) / self.tau)  # N
        # softmax
        softmax = D2C_sim / (D2C_sim + D2D_sim + 1e-8)
        return softmax`

**Comment by MINGUKKANG** [comment]
Hello,

Thank you for your interest in our work:)

It seems to be right except for the absence of lines for numerical stability.

Also, if you want to use Reboot_Softmax as a training objective, applying "-tf.math.reduce_mean(log(*))"  is required. 

Thank you.

Best,

Minguk

**Comment by tom99763** [comment]
> Hello,
> 
> Thank you for your interest in our work:)
> 
> It seems to be right except for the absence of lines for numerical stability.
> 
> Also, if you want to use Reboot_Softmax as a training objective, applying "-tf.math.reduce_mean(log(*))" is required.
> 
> Thank you.
> 
> Best,
> 
> Minguk

Thank you , the stablization method you use is to shift all the values by the maximum value along second dimension? 

And I have a quesion, in your implement of Data2DataCrossEntropyLoss:

`neg_repul=torch.log(torch.exp(-pos_attr))+....)` 

I confuse in this term, why you add minus to pos_attr, this is different to the formulation in the paper.

**Comment by MINGUKKANG** [comment]
Thank you , the stablization method you use is to shift all the values by the maximum value along second dimension?
=> Yes, this method is also used in the supervised contrastive learning github.

```python
sim_matrix = self.remove_diag(sim_matrix/self.temperature)
# for numerical stability
sim_max, _ = torch.max(sim_matrix, dim=1, keepdim=True)
 sim_matrix = F.relu(sim_matrix) - sim_max.detach()
```

```
neg_repul=torch.log(torch.exp(-pos_attr))+....)
```
I confuse in this term, why you add minus to pos_attr, this is different to the formulation in the paper.
=> This is because our implementation uses ReLU function instead of the minimum operation []_{-}. Thus, the sign should be reverse.

If you decompose D2D-CE loss using the log property log(A/B) = log(A) - log(B), you might understand our implementation more clearly.

**Comment by tom99763** [comment]
> Thank you , the stablization method you use is to shift all the values by the maximum value along second dimension? => Yes, this method is also used in the supervised contrastive learning github.
> 
> ```python
> sim_matrix = self.remove_diag(sim_matrix/self.temperature)
> # for numerical stability
> sim_max, _ = torch.max(sim_matrix, dim=1, keepdim=True)
> sim_matrix = F.relu(sim_matrix) - sim_max.detach()
> ```
> 
> ```
> neg_repul=torch.log(torch.exp(-pos_attr))+....)
> ```
> 
> I confuse in this term, why you add minus to pos_attr, this is different to the formulation in the paper. => This is because our implementation uses ReLU function instead of the minimum operation []_{-}. Thus, the sign should be reverse.
> 
> If you decompose D2D-CE loss using the log property: log(A/B) = log(A) - log(B), you might understand our implementation more clearly.

Thank you, I understand !
